storage/storage_image.go

```diff
@@ -476,14 +550,175 @@ func (s *storageImageDestination) getConfigBlob(info types.BlobInfo) ([]byte, er
+	}
+	// Compute the manifest digest.
+	options := &storage.ImageOptions{}
+	if manifestDigest, err := manifest.Digest(m); err == nil {
+		options.Digest = manifestDigest
+	}
+	// Create the image record, pointing to no layers.
+	intendedID := s.imageRef.id
+	if instanceDigest != nil {
+		intendedID = ""
+	}
+	if intendedID == "" {
+		intendedID = man.ImageID()
+	}
+	return s.commitImageRecord(instanceDigest, m, intendedID, "", options)
+}
+
+func (s *storageImageDestination) commitImageRecord(instanceDigest *digest.Digest, manifest []byte, intendedID string, layer string, options *storage.ImageOptions) error {
+	oldNames := []string{}
+	img, err := s.imageRef.transport.store.CreateImage(intendedID, nil, layer, "", options)
 	if err != nil {
-		return errors.Wrapf(err, "error parsing manifest")
+		if errors.Cause(err) != storage.ErrDuplicateID {
+			logrus.Debugf("error creating image: %q", err)
+			return errors.Wrapf(err, "error creating image %q", intendedID)
+		}
+		img, err = s.imageRef.transport.store.Image(intendedID)
+		if err != nil {
+			return errors.Wrapf(err, "error reading image %q", intendedID)
+		}
+		if img.TopLayer != layer {
+			why := "uses layers"
+			if layer != "" {
+				why = "uses a different top layer"
+			}
+			logrus.Debugf("error creating image: image with ID %q exists, but %s", intendedID, why)
+			return errors.Wrapf(storage.ErrDuplicateID, "image with ID %q already exists, but %s", intendedID, why)
+		}
+		logrus.Debugf("reusing image ID %q", img.ID)
+		oldNames = append(oldNames, img.Names...)
+	} else {
+		logrus.Debugf("created new image ID %q", img.ID)
+	}
+	// Add the non-layer blobs as data items.  Since we only share layers, they should all be in files, so
+	// we just need to screen out the ones that are actually layers to get the list of non-layers.
+	if err = s.commitDataBlobs(img.ID); err != nil {
+		if _, err2 := s.imageRef.transport.store.DeleteImage(img.ID, true); err2 != nil {
+			logrus.Debugf("error deleting incomplete image %q: %v", img.ID, err2)
+		}
+		return errors.Wrapf(err, "error saving non-layer blobs to image %q", img.ID)
+	}
+	// Set the reference's name on the image.
+	if err = s.commitName(img.ID, oldNames, instanceDigest); err != nil {
+		if _, err2 := s.imageRef.transport.store.DeleteImage(img.ID, true); err2 != nil {
+			logrus.Debugf("error deleting incomplete image %q: %v", img.ID, err2)
+		}
+		return errors.Wrapf(err, "error setting name on image %q", img.ID)
+	}
+	// Save the manifest.  Use storage.ImageDigestBigDataKey as the item's
+	// name, so that its digest can be used to locate the image in the Store.
+	if err := s.imageRef.transport.store.SetImageBigData(img.ID, storage.ImageDigestBigDataKey, manifest); err != nil {
+		if _, err2 := s.imageRef.transport.store.DeleteImage(img.ID, true); err2 != nil {
+			logrus.Debugf("error deleting incomplete image %q: %v", img.ID, err2)
+		}
+		logrus.Debugf("error saving manifest for image %q: %v", img.ID, err)
+		return err
```

Comment:
The errors prior to this you wrapped with a little verbiage.  This one and below you did a debug then straight error.  I'd vote continuing on with the error wrap using the veribage from the debugs, but your call.

Reply:
Changed them as you described.